### PR TITLE
NEW Management of retained warranty on situation invoices

### PR DESCRIPTION
**Pull request by john-botella**
\+ New admin page for situation invoices ans retained warranty : display if hidden conf for situation invoice is activated
\+ New retained warranty for situation invoices (hidden conf). (note: next step is to allow standard invoices to use retained warranty)

For info at ATM we use retained warranty from Dolibarr 8.0 (branch 8.0_BTP, 9.0_BTP)

### CODE REVIEW

File: htdocs/admin/facture_situation.php [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+print '</form>';
+
+dol_fiche_end();
+
+// End of page
+llxFooter();
+$db->close();
+
+function _updateBtn(){
```
Comment (eldy) [inline htdocs/admin/facture_situation.php]: Functions must have code comments.

File: htdocs/admin/facture_situation.php [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    global $langs;
+    print '<div style="text-align: right;" >';
+    print '<input type="submit" class="butAction" value="'.$langs->trans("Save").'">';
+    print '</div>';
+}
+
+function _print_on_off($confkey, $title = false, $desc ='')
```
Comment (eldy) [inline htdocs/admin/facture_situation.php]: Functions must have code comments.

File: htdocs/admin/facture_situation.php [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+    print '</td>';
+    print '<td align="center" width="20">&nbsp;</td>';
+    print '<td align="right" width="300">';
+    print ajax_constantonoff($confkey);
+    print '</td></tr>';
+}
+
+
+function _print_input_form_part($confkey, $title = false, $desc ='', $metas = array(), $type='input', $help = false)
```
Comment (eldy) [inline htdocs/admin/facture_situation.php]: Functions must have code comments.

**Comment by eldy** [comment]
Can you explain how the retain warranty works ?
Can you also add a hidden option to enable this ?

**Comment by john-botella** [comment]
Retainage is a term of art in the construction industry. It consists of contract funds earned by a contractor for work performed on a project that the owner is entitled by contract to withhold from payments to the contractor until some specified time or event occurs to trigger the owner’s obligation to pay the contractor some or all of the retained funds. The amount of retainage the owner is entitled to retain is typically defined in the construction contract as a percentage of progress payments periodically due for completed work. Conditions to the contractor’s entitlement to payment of the retained funds are likewise typically defined in the contract, being usually tied to the occurrence of a particular event(s) (e.g. substantial completion) or stage of the project (e.g. when the work is 50% complete). Contractors constrained by such retainage provisions in their contract with the owner will typically reserve the right to withhold funds earned by their subcontractors in the same, or in some cases greater percentage as the owner is permitted to withhold funds from the contractor. In all cases, however, the purpose of retaining funds earned under the contract is to protect one contracting party from the inability or unwillingness of the other contracting party to remedy defective work, discharge liens filed against the property, or otherwise comply with its obligations under the contract.

**Résumé rapide en français:**
La retenue de garantie est une somme d'argent représentant x% du montant total des travaux. C'est un vocable utilisé lors d'une commande publique correspondant à des marchés de travaux. La retenue de garantie fait référence à la loi du 16 juillet 1971 en France. Si le maître d'ouvrage (le propriétaire) ou le maître d'oeuvre (la personne qui a organisé et contrôlé les travaux) constate des défauts ou des malfaçons, l'entreprise ayant réalisé les travaux se voit retenir une somme égale à x% du montant des travaux. La retenue de cette somme permet au maître d'ouvrage d'exercer une pression sur l'entreprise jusqu'à exécution complète et correcte des travaux.

**Comment by john-botella** [comment]
I have added a admin page who can activate retained warranty but this page is already hidden by an hidden conf for situation invoices. Does I need to really hide this conf ? (INVOICE_USE_SITUATION_RETAINED_WARRANTY)

**Comment by eldy** [comment]
Sure, but how does this work ? What is process ?

**Comment by eldy** [comment]
> I have added a admin page who can activate retained warranty but this page is already hidden by an hidden conf for situation invoices. Does I need to really hide this conf ? (INVOICE_USE_SITUATION_RETAINED_WARRANTY)

No, it's enough

**Comment by john-botella** [comment]
**EN**
Currently they work exclusively for situation invoices (but we plan to allow the use on any type of invoice )

Retention of warranty work in the following ways:
A percentage of the retention  and a deadline for payment of the retention are defined to an invoice.
This information appears on the PDF.
The bill then awaits 2 payments. 
The first corresponding to the total With taxes - the percentage of the retention of warranty, the second corresponds to the remaining amount equivalent to the retention of warranty.

The invoice display an unpaid alert according the retainage deadlines too.



-------------------------
**FR**
Actuellement elles fonctionnent exclusivement pour les factures de situation

Les retenues de garanties fonctionnement de la manières suivantes:
on défini un percentage de retenu de garantie appliquée à une facture et la date limite de paiement de la retenu de garantie.
Ces informations apparaisses sur sur le PDF.
La facture attend alors 2 règlements. Le premier correspondant au total TTC - le pourcentage de retenu de garantie, le deuxième correspond au montant restant equivalent à la retenu de garantie.

La facture affiche une alerte impayé en prenant en compte la date limite de règlement de la retenu de garantie.

**Comment by john-botella** [comment]
@eldy : you can remove the "PR to fix - Continuous Integration Test failed" label

**Pull request merged**